includes/SVG.php

```diff
@@ -0,0 +1,428 @@
+				$view_box = $xml->getAttribute( 'viewbox' );
+			}
+			$pieces = explode( ' ', $view_box );
+			if ( 4 === count( $pieces ) ) {
+				list (, , $width, $height ) = $pieces;
+			}
+		}
+
+		if ( ! $width || ! $height ) {
+			return new WP_Error( 'invalid_svg_size', __( 'Unable to generate SVG image size.', 'web-stories' ) );
+		}
+
+		return compact( 'width', 'height' );
+	}
+
+	/**
+	 * Sanitize the SVG
+	 *
+	 * @since 1.2.0
+	 *
+	 * @param string $file File path.
+	 *
+	 * @return true|WP_Error
+	 */
+	protected function sanitize( $file ) {
+		$dirty     = $this->get_svg_data( $file );
+		$sanitizer = new Sanitizer();
+		$clean     = $sanitizer->sanitize( $dirty );
+
+		if ( empty( $clean ) ) {
+			return new WP_Error( 'invalid_xml_svg', __( 'Invalid xml in SVG.', 'web-stories' ) );
+		}
+
+		$errors = $sanitizer->getXmlIssues();
+		if ( count( $errors ) > 1 ) {
+			return new WP_Error( 'insecure_svg_file', __( "Sorry, this file couldn't be sanitized so for security reasons wasn't uploaded.", 'web-stories' ) );
+		}
+
+		return true;
+	}
+
+	/**
+	 * Work around for incorrect mime type.
+	 *
+	 * @since 1.2.0
+	 *
+	 * @param array       $wp_check_filetype_and_ext {
+	 *                                               Values for the extension, mime type, and corrected filename.
+	 *
+	 * @type string|false $ext                       File extension, or false if the file doesn't match a mime type.
+	 * @type string|false $type                      File mime type, or false if the file doesn't match a mime type.
+	 * @type string|false $proper_filename           File name with its correct extension, or false if it cannot be
+	 *       determined.
+	 * }
+	 *
+	 * @param string      $file                      Full path to the file.
+	 * @param string      $filename                  The name of the file (may differ from $file due to
+	 *                                               $file being in a tmp directory).
+	 * @param string[]    $mimes                     Array of mime types keyed by their file extension regex.
+	 * @param string|bool $real_mime                 The actual mime type or false if the type cannot be determined.
+	 *
+	 * @return array
+	 */
+	public function wp_check_filetype_and_ext( $wp_check_filetype_and_ext, $file, $filename, $mimes, $real_mime ) {
+		if ( 'image/svg' === $real_mime ) {
+			$wp_check_filetype_and_ext = [
+				'ext'             => self::EXT,
+				'type'            => self::MIME_TYPE,
+				'proper_filename' => false,
+			];
+		}
+
+		return $wp_check_filetype_and_ext;
+	}
+
+	/**
+	 * Get xml document.
+	 *
+	 * @since 1.2.0
+	 *
+	 * @param string $svg String of xml.
+	 *
+	 * @return DOMElement|false
+	 */
+	protected function get_xml( $svg ) {
+		$dom                      = new DOMDocument();
+		$dom->preserveWhiteSpace  = false;
+		$dom->strictErrorChecking = false;
+
+		$errors = libxml_use_internal_errors( true );
+		$loaded = $dom->loadXML( $svg );
+		if ( ! $loaded ) {
+			return false;
```

Comment:
There's no test coverage for this line. Are we testing with an invalid XML file?

Reply:
Fixed https://github.com/google/web-stories-wp/pull/5103/commits/7b55319d7d30890e2af1e0c2bab44c72af152e44